api/src/main/java/com/inrupt/client/ProblemDetails.java

```diff
@@ -0,0 +1,143 @@
+import java.net.URI;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Optional;
+
+/**
+ * A data class representing a structured problem description sent by the server on error response.
+ *
+ * @see <a href="https://www.rfc-editor.org/rfc/rfc9457">RFC 9457 Problem Details for HTTP APIs</a>
+ */
+public class ProblemDetails {
+    public static final String MIME_TYPE = "application/problem+json";
+    public static final String DEFAULT_TYPE = "about:blank";
+    private final URI type;
+    private final String title;
+    private final String details;
+    private final int status;
+    private final URI instance;
+    private static JsonService jsonService;
+    private static boolean isJsonServiceInitialized;
+
+    public ProblemDetails(
+        final URI type,
+        final String title,
+        final String details,
+        final int status,
+        final URI instance
+    ) {
+        this.type = type;
+        this.title = title;
+        this.details = details;
+        this.status = status;
+        this.instance = instance;
+    }
+
+    public URI getType() {
+        return this.type;
+    };
+
+    public String getTitle() {
+        return this.title;
+    };
+
+    public String getDetails() {
+        return this.details;
+    };
+
+    public int getStatus() {
+        return this.status;
+    };
+
+    public URI getInstance() {
+        return this.instance;
+    };
+
+    private static JsonService getJsonService() {
+        if (ProblemDetails.isJsonServiceInitialized) {
+            return ProblemDetails.jsonService;
+        }
+        // It is acceptable for a JenaBodyHandlers instance to be in a classpath without any implementation for
+        // JsonService, in which case the ProblemDetails exceptions will fallback to default and not be parsed.
+        try {
+            ProblemDetails.jsonService = ServiceProvider.getJsonService();
+        } catch (IllegalStateException e) {
+            ProblemDetails.jsonService = null;
+        }
+        ProblemDetails.isJsonServiceInitialized = true;
+        return ProblemDetails.jsonService;
+    }
+
+    public static ProblemDetails fromErrorResponse(
+            final int statusCode,
+            final Headers headers,
+            final byte[] body
+    ) {
+        final JsonService jsonService = getJsonService();
+        if (jsonService == null
+                || (headers != null && !headers.allValues("Content-Type").contains(ProblemDetails.MIME_TYPE))) {
+            return new ProblemDetails(
+                URI.create(ProblemDetails.DEFAULT_TYPE),
+                HttpStatus.StatusMessages.getStatusMessage(statusCode),
+                null,
+                statusCode,
+                null
+            );
+        }
+        try {
+            final ProblemDetailsData pdData = jsonService.fromJson(
+                    new ByteArrayInputStream(body),
+                    ProblemDetailsData.class
+            );
```

Comment:
I tested this with both jackson and jsonb. The test coverage is added in a different PR.

Reply:
That turned out to be a lie.